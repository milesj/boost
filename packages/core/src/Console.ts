/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import exit from 'exit';
import cliSize from 'term-size';
import ansiEscapes from 'ansi-escapes';
import Emitter from './Emitter';
import Tool from './Tool';
import { Debugger } from './types';
import Output from './Output';

// 16 FPS (60 FPS is actually too fast as it tears)
export const FPS_RATE = 62.5;

// Bind our writable streams for easy access
export const BOUND_WRITERS = {
  stderr: process.stderr.write.bind(process.stderr),
  stdout: process.stdout.write.bind(process.stdout),
};

export const WRAPPED_STREAMS = {
  stderr: false,
  stdout: false,
};

export type StreamType = 'stderr' | 'stdout';

export interface ConsoleState {
  disabled: boolean;
  final: boolean;
  started: boolean;
  stopped: boolean;
}

export default class Console extends Emitter {
  bufferedStreams: (() => void)[] = [];

  debug: Debugger;

  errorLogs: string[] = [];

  logs: string[] = [];

  outputQueue: Output[] = [];

  tool: Tool<any>;

  protected renderTimer: NodeJS.Timer | null = null;

  protected restoreCursorOnExit: boolean = false;

  protected state: ConsoleState = {
    disabled: false,
    final: false,
    started: false,
    stopped: false,
  };

  private writers: typeof BOUND_WRITERS;

  constructor(tool: Tool<any>, /* test only */ testWriters: typeof BOUND_WRITERS = BOUND_WRITERS) {
    super();

    this.debug = tool.createDebugger('console');
    this.tool = tool;
    this.writers = testWriters;

    // istanbul ignore next
    if (process.env.NODE_ENV !== 'test') {
      process
        .on('SIGINT', this.handleSignal)
        .on('SIGTERM', this.handleSignal)
        .on('uncaughtException', this.handleFailure)
        .on('unhandledRejection', this.handleFailure);
    }
  }

  /**
   * Disable the render loop entirely.
   */
  disable(): this {
    this.state.disabled = true;

    return this;
  }

  /**
   * Display a footer after all output.
   */
  displayFooter() {
    const { footer } = this.tool.options;

    if (footer) {
      this.out(footer, 1);
    }
  }

  /**
   * Display a header before all output.
   */
  displayHeader() {
    const { header } = this.tool.options;

    if (header) {
      this.out(header, 1);
    }
  }

  /**
   * Enable the render loop.
   */
  enable(): this {
    this.state.disabled = false;

    return this;
  }

  /**
   * Write a message to `stderr` with optional trailing newline(s).
   */
  err(message: string, nl: number = 0): this {
    if (!this.isSilent()) {
      this.writers.stderr(message + '\n'.repeat(nl));
    }

    return this;
  }

  /**
   * Flush buffered stream output.
   */
  flushBufferedStreams(): this {
    this.bufferedStreams.forEach(buffer => {
      buffer();
    });

    return this;
  }

  /**
   * Flush the top output block in the queue.
   */
  flushOutputQueue(): this {
    const outputs = this.outputQueue.filter((out, i) => i === 0 || out.isConcurrent());

    // Erase the previous output
    outputs.forEach(output => {
      output.erasePrevious();
    });

    // Write buffered output
    this.flushBufferedStreams();

    // Write the next output
    outputs.forEach(output => {
      output.render();
    });

    // Remove completed outputs
    this.outputQueue = this.outputQueue.filter(out => !out.isComplete());

    // Stop the render loop once the queue is empty
    if (this.isEmptyQueue()) {
      this.stopRenderLoop();
    } else {
      this.startRenderLoop();
    }

    return this;
  }

  /**
   * Handle uncaught exceptions and unhandled rejections that bubble up.
   */
  handleFailure = (error: Error) => {
    this.start();
    this.debug('Uncaught exception or unresolved promise handled');
    this.stop(error);

    exit(2);
  };

  /**
   * Handle SIGINT and SIGTERM interruptions.
   */
  handleSignal = () => {
    this.start();
    this.debug('SIGINT or SIGTERM handled');
    this.stop(new Error(this.tool.msg('errors:processTerminated')));

    exit(2);
  };

  /**
   * Hide the console cursor.
   */
  hideCursor(): this {
    this.out(ansiEscapes.cursorHide);

    if (!this.restoreCursorOnExit && !this.isSilent()) {
      this.restoreCursorOnExit = true;

      // istanbul ignore next
      process.on('exit', () => {
        process.stdout.write(ansiEscapes.cursorShow);
      });
    }

    return this;
  }

  /**
   * Return true if the render loop has been disabled.
   */
  isDisabled(): boolean {
    return this.state.disabled;
  }

  /**
   * Return true if the output queue is empty.
   */
  isEmptyQueue(): boolean {
    return this.outputQueue.length === 0;
  }

  /**
   * Return true if the final render.
   */
  isFinalRender(): boolean {
    return this.state.final;
  }

  /**
   * Return true if the there should be no output.
   */
  isSilent(): boolean {
    return this.tool.config.silent;
  }

  /**
   * Return true if the defined stream has been wrapped by the console layer.
   */
  isStreamWrapped(type: StreamType): boolean {
    return !!WRAPPED_STREAMS[type];
  }

  /**
   * Log a message to display on success during the final render.
   */
  log(message: string): this {
    this.logs.push(message);

    return this;
  }

  /**
   * Log a live message to display during the rendering process.
   */
  logLive(message: string): this {
    // Write to the wrapped buffer
    process.stdout.write(message);

    return this;
  }

  /**
   * Log an error message to display on failure during the final render.
   */
  logError(message: string): this {
    this.errorLogs.push(message);

    return this;
  }

  /**
   * Write a message to `stdout` with optional trailing newline(s).
   */
  out(message: string, nl: number = 0): this {
    if (!this.isSilent()) {
      this.writers.stdout(message + '\n'.repeat(nl));
    }

    return this;
  }

  /**
   * Enqueue a block of output to be rendered.
   */
  render(output: Output): this {
    if (this.isDisabled()) {
      throw new Error(
        'Output cannot be enqueued as the render loop has been disabled. This is usually caused by conflicting reporters.',
      );
    }

    if (!this.outputQueue.includes(output)) {
      this.outputQueue.push(output);
    }

    // Only run the render loop when output is enqueued
    this.startRenderLoop();

    return this;
  }

  /**
   * Handle the final rendering of all output before stopping.
   */
  renderFinalOutput(error: Error | null) {
    this.debug('Rendering final console output');
    this.state.final = true;

    // Mark all output as final
    this.outputQueue.forEach(output => {
      output.enqueue(true);
    });

    // Recursively render the remaining output
    while (!this.isEmptyQueue()) {
      this.flushOutputQueue();
    }

    // Stop the render loop
    this.stopRenderLoop();

    if (error) {
      if (this.errorLogs.length > 0) {
        this.err(`\n${this.errorLogs.join('\n')}\n`);
      }

      this.emit('error', [error]);
    } else {
      if (this.logs.length > 0) {
        this.out(`\n${this.logs.join('\n')}\n`);
      }

      this.displayFooter();
    }

    // Flush any stream output that still exists
    this.flushBufferedStreams();

    // Show the cursor incase it has been hidden
    this.out(ansiEscapes.cursorShow);
  }

  /**
   * Reset the cursor back to the bottom of the console.
   */
  resetCursor(): this {
    this.out(ansiEscapes.cursorTo(0, cliSize().rows));

    return this;
  }

  /**
   * Show the console cursor.
   */
  showCursor(): this {
    this.restoreCursorOnExit = false;
    this.out(ansiEscapes.cursorShow);

    return this;
  }

  /**
   * Start the console by wrapping streams and buffering output.
   */
  start(args: any[] = []): this {
    if (this.state.started) {
      return this;
    }

    this.debug('Starting console render loop');
    this.emit('start', args);
    this.wrapStreams();
    this.displayHeader();
    this.state.started = true;

    return this;
  }

  /**
   * Automatically render the console in a timeout loop at 16 FPS.
   */
  startRenderLoop() {
    if (this.isSilent() || this.isDisabled()) {
      return;
    }

    this.renderTimer = setTimeout(() => {
      this.flushOutputQueue();
    }, FPS_RATE);
  }

  /**
   * Stop the console rendering process.
   */
  stop(error: Error | null = null) {
    if (this.state.stopped) {
      return;
    }

    if (error) {
      this.debug('Stopping console with an error');
      this.errorLogs.push(...this.logs);
      this.logs = [];
    } else {
      this.debug('Stopping console render loop');
    }

    this.renderFinalOutput(error);
    this.unwrapStreams();
    this.emit('stop', [error]);
    this.state.stopped = true;
    this.state.started = false;
  }

  /**
   * Stop the background render loop.
   */
  stopRenderLoop() {
    if (this.renderTimer) {
      clearTimeout(this.renderTimer);
      this.renderTimer = null;
    }
  }

  /**
   * Unwrap the native console and reset it back to normal.
   */
  unwrapStreams() {
    if (this.isSilent() || process.env.NODE_ENV === 'test') {
      return;
    }

    ['stderr', 'stdout'].forEach(key => {
      const name = key as StreamType;

      if (!this.isStreamWrapped(name)) {
        return;
      }

      this.debug('Unwrapping `%s` stream', name);

      process[name].write = this.writers[name] as any;

      WRAPPED_STREAMS[name] = false;
    });
  }

  /**
   * Wrap the `stdout` and `stderr` streams and buffer the output as
   * to not collide with our render loop.
   */
  wrapStreams() {
    if (this.isSilent() || this.isDisabled() || process.env.NODE_ENV === 'test') {
      return;
    }

    ['stderr', 'stdout'].forEach(key => {
      const name = key as StreamType;
      const stream = process[name];
      let buffer = '';

      if (this.isStreamWrapped(name)) {
        return;
      }

      this.debug('Wrapping `%s` stream', name);

      this.bufferedStreams.push(() => {
        if (stream.isTTY && buffer) {
          this.writers[name](buffer);
        }

        buffer = '';
      });

      stream.write = (chunk: string) => {
        // No output, display immediately
        if (this.isEmptyQueue()) {
          this.writers[name](chunk);
        } else {
          buffer += String(chunk);
        }

        return true;
      };

      WRAPPED_STREAMS[name] = true;
    });
  }
}
