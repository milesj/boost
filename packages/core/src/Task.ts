/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Context from './Context';
import {
  STATUS_PENDING,
  STATUS_RUNNING,
  STATUS_SKIPPED,
  STATUS_PASSED,
  STATUS_FAILED,
} from './constants';
import { Status } from './types';
import Emitter from './Emitter';

export type TaskAction<Ctx extends Context> = (
  context: Ctx,
  value: any,
  task: Task<Ctx>,
) => any | Promise<any>;

export interface TaskMetadata {
  depth: number;
  index: number;
  startTime: number;
  stopTime: number;
}

export default class Task<Ctx extends Context> extends Emitter {
  action: TaskAction<Ctx>;

  // @ts-ignore Set after instantiation
  context: Ctx;

  title: string;

  metadata: TaskMetadata = {
    depth: 0,
    index: 0,
    startTime: 0,
    stopTime: 0,
  };

  output: string = '';

  parent: Task<Ctx> | null = null;

  status: Status = STATUS_PENDING;

  statusText: string = '';

  tasks: Task<Ctx>[] = [];

  constructor(title: string, action: TaskAction<Ctx>) {
    super();

    if (!title || typeof title !== 'string') {
      throw new Error('Tasks require a title.');
    }

    if (action !== null && typeof action !== 'function') {
      throw new Error('Tasks require an executable function.');
    }

    this.action = action;
    this.status = STATUS_PENDING;
    this.title = title;
  }

  /**
   * Return true if the task failed when executing.
   */
  hasFailed(): boolean {
    return this.status === STATUS_FAILED;
  }

  /**
   * Return true if the task executed successfully.
   */
  hasPassed(): boolean {
    return this.status === STATUS_PASSED;
  }

  /**
   * Return true if the task has been completed in any form.
   */
  isComplete(): boolean {
    return this.hasPassed() || this.hasFailed() || this.isSkipped();
  }

  /**
   * Return true if the task has not been executed yet.
   */
  isPending(): boolean {
    return this.status === STATUS_PENDING;
  }

  /**
   * Return true if the task is currently running.
   */
  isRunning(): boolean {
    return this.status === STATUS_RUNNING;
  }

  /**
   * Return true if the task was or will be skipped.
   */
  isSkipped(): boolean {
    return this.status === STATUS_SKIPPED;
  }

  /**
   * Run the current task by executing it and performing any before and after processes.
   */
  async run(context: Ctx, value: any): Promise<any> {
    this.setContext(context);
    this.emit('run', [value]);

    if (this.isSkipped()) {
      this.status = STATUS_SKIPPED;
      this.emit('skip', [value]);

      return Promise.resolve(value);
    }

    this.status = STATUS_RUNNING;
    this.metadata.startTime = Date.now();

    try {
      this.output = await this.action(context, value, this);
      this.status = STATUS_PASSED;
      this.metadata.stopTime = Date.now();
      this.emit('pass', [this.output]);
    } catch (error) {
      this.status = STATUS_FAILED;
      this.metadata.stopTime = Date.now();
      this.emit('fail', [error]);

      throw error;
    }

    this.statusText = '';

    return this.output;
  }

  /**
   * Set the context to be passed around.
   */
  setContext(context: Ctx): this {
    this.context = context;

    return this;
  }

  /**
   * Mark a task as skipped if the condition is true.
   */
  skip(condition: boolean = true): this {
    if (condition) {
      this.status = STATUS_SKIPPED;
    }

    return this;
  }
}
