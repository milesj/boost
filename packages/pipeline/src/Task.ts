import { Blueprint, Predicates } from '@boost/common';
import WorkUnit from './WorkUnit';

export default class Task<Input = unknown, Output = Input> extends WorkUnit<{}, Input, Output> {
	// A task is simply a work unit that executes a standard function.
	// It doesn't need configurable options, so implement an empty blueprint.
	blueprint(preds: Predicates): Blueprint<object> {
		return {};
	}
}
