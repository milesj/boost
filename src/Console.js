/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import chalk from 'chalk';
import ExitError from './ExitError';

import type Task from './Task';
import type Reporter from './Reporter';
import type { ToolConfig, ToolOptions } from './types';

const DEBUG_COLORS: string[] = [
  'white',
  'cyan',
  'blue',
  'magenta',
  'red',
  'yellow',
  'green',
];

export default class Console<Tr: Reporter<Object>> {
  debugs: string[] = [];

  debugGroups: string[] = [];

  debugIndex: number = -1;

  errors: string[] = [];

  interrupted: boolean = false;

  logs: string[] = [];

  reporter: Tr;

  constructor(reporter: Tr) {
    this.reporter = reporter;

    // Avoid binding listeners while testing
    if (process.env.NODE_ENV === 'test') {
      return;
    }

    const signalHandler = () => {
      if (this.interrupted) {
        this.exit('Process has been terminated.');
      } else {
        this.log(chalk.yellow('Press Ctrl+C again to exit.'));
        this.interrupted = true;
      }
    };

    process
      .on('SIGINT', signalHandler)
      .on('SIGTERM', signalHandler)
      .on('uncaughtException', (error) => {
        this.log(chalk.yellow('Uncaught exception detected!'));
        this.exit(error.message);
      })
      .on('unhandledRejection', (error) => {
        this.log(chalk.yellow('Unhandled promise rejection detected!'));
        this.exit(error.message);
      });
  }

  /**
   * Add a message to the debug log.
   */
  debug(message: string) {
    this.debugs.push(`${chalk.gray('[debug]')}${this.reporter.indent(this.debugGroups.length)} ${message}`);
  }

  /**
   * Add a message to the error log.
   */
  error(message: string) {
    this.errors.push(message);
  }

  /**
   * Force exit the application.
   */
  /* istanbul ignore next */
  exit(message: string | Error | null, code?: number = 1) {
    let errorCode = code;

    // Null messages are always successful
    if (message !== null) {
      const error = (message instanceof Error) ? message : new ExitError(message, code);

      if (typeof error.code === 'number') {
        errorCode = error.code;
      }
    }

    // Show terminal cursor
    this.log('\x1B[?25h'); // eslint-disable-line unicorn/no-hex-escape

    // Stop the renderer
    this.stop();

    // Send final output and exit after buffer is flushed
    (code === 0 ? process.stdout : process.stderr).write(this.reporter.render(code), () => {
      process.exit(errorCode); // eslint-disable-line unicorn/no-process-exit
    });
  }

  /**
   * Add a message to the output log.
   */
  log(message: string) {
    this.logs.push(message);
  }

  /**
   * Start a new console and begin rendering.
   */
  start(config: ToolConfig, options: ToolOptions, tasks: Task<*, *>[]) {
    const { debug, silent } = config;
    const { footer, header } = options;
    const { debugs, errors, logs } = this;

    this.reporter.start(() => ({
      debug,
      debugs,
      errors,
      footer,
      header,
      logs,
      silent,
      tasks,
    }));
  }

  /**
   * Start a debug capturing group, which will indent all incoming debug messages.
   */
  startDebugGroup(group: string) {
    this.debugIndex += 1;

    if (this.debugIndex === DEBUG_COLORS.length) {
      this.debugIndex = 0;
    }

    const color = DEBUG_COLORS[this.debugIndex];

    this.debug(chalk[color](`[${group}]`));
    this.debugGroups.push(group);
  }

  /**
   * Stop rendering and end the console.
   */
  stop() {
    this.reporter.stop();
  }

  /**
   * End the current debug capturing group.
   */
  stopDebugGroup() {
    const color = DEBUG_COLORS[this.debugIndex];
    const group = this.debugGroups.pop();

    this.debug(chalk[color](`[/${group}]`));
    this.debugIndex -= 1;

    if (this.debugIndex < 0) {
      this.debugIndex = DEBUG_COLORS.length - 1;
    }
  }

  /**
   * Force a rendering update.
   */
  update() {
    this.reporter.update();
  }
}
