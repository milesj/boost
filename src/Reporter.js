/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import chalk from 'chalk';
import figures from 'figures';
import logUpdate from 'log-update';
import Options, { bool, number } from 'optimal';
import {
  STATUS_PENDING,
  STATUS_RUNNING,
  STATUS_SKIPPED,
  STATUS_PASSED,
  STATUS_FAILED,
} from './constants';

import type Task from './Task';
import type { ReportLoader, ReporterOptions } from './types';

export default class Reporter<Tx: Object> {
  instance: number = 0;

  loader: ?ReportLoader<Tx> = null;

  moduleName: string = '';

  name: string = '';

  options: ReporterOptions;

  constructor(options?: Object = {}) {
    this.options = new Options(options, {
      refreshRate: number(100), // eslint-disable-line no-magic-numbers
      silent: bool(),
    }, {
      name: 'Reporter',
    });
  }

  /**
   * Create an indentation based on the defined length.
   */
  indent(length: number): string {
    return '  '.repeat(length);
  }

  /**
   * Render the output by looping over all tasks and messages.
   */
  render(code?: number = 0): string {
    if (!this.loader) {
      return '';
    }

    const output = [];
    const {
      debugs = [],
      errors = [],
      logs = [],
      tasks = [],
    } = this.loader();

    // Tasks first
    tasks.forEach((task) => {
      output.push(...this.renderTask(task, 0));
    });

    // Debugs second
    debugs.forEach((debug) => {
      output.push(debug);
    });

    // Logs last
    (code === 0 ? logs : errors).forEach((log) => {
      output.push(log);
    });

    return output.join('\n');
  }

  /**
   * Render a single task including its title and status.
   * If sub-tasks or sub-routines exist, render them recursively.
   */
  renderTask(task: Task<*, Tx>, level?: number = 0, suffix?: string = ''): string[] {
    const output = [];

    // Generate the message row
    let message = `${this.indent(level)}${this.renderStatus(task)} ${task.title}`;

    if (task.isSkipped()) {
      message += ` ${chalk.yellow('[skipped]')}`;

    } else if (task.hasFailed()) {
      message += ` ${chalk.red('[failed]')}`;

    } else if (suffix) {
      message += ` ${suffix}`;
    }

    output.push(message);

    // Show only one sub-task at a time
    if (task.subtasks.length > 0) {
      let pendingTask;
      let runningTask;
      let failedTask;
      let passed = 0;

      task.subtasks.forEach((subTask) => {
        if (subTask.isPending() && !pendingTask) {
          pendingTask = subTask;

        } else if (subTask.isRunning() && !runningTask) {
          runningTask = subTask;

        } else if (subTask.hasFailed() && !failedTask) {
          failedTask = subTask;

        } else if (subTask.hasPassed()) {
          passed += 1;
        }
      });

      const activeTask = failedTask || runningTask || pendingTask;
      const taskSuffix = chalk.gray(`[${passed}/${task.subtasks.length}]`);

      // Only show if the parent is running or a task failed
      if (activeTask && (task.isRunning() || failedTask)) {
        output.push(...this.renderTask(activeTask, level + 1, taskSuffix));
      }
    }

    // Show the current status output
    if (task.statusText) {
      output.push(`${this.indent(level + 1)}${chalk.gray(task.statusText)}`);
    }

    // Show all sub-routines
    if (task.subroutines.length > 0) {
      task.subroutines.forEach((routine) => {
        output.push(...this.renderTask(routine, level + 1));
      });
    }

    return output;
  }

  /**
   * Render a status symbol for a task.
   */
  renderStatus(task: Task<*, Tx>): string {
    switch (task.status) {
      case STATUS_PENDING:
        return chalk.gray(figures.bullet);
      case STATUS_RUNNING:
        return chalk.gray(task.spinner());
      case STATUS_SKIPPED:
        return chalk.yellow(figures.circleDotted);
      case STATUS_PASSED:
        return chalk.green(figures.tick);
      case STATUS_FAILED:
        return chalk.red(figures.cross);
      default:
        return '';
    }
  }

  /**
   * Start the output process once setting the task loader.
   */
  start(loader: ReportLoader<Tx>) {
    if (this.instance) {
      return;
    } else if (!loader || typeof loader !== 'function') {
      throw new TypeError('A loader is required to render console output.');
    }

    this.loader = loader;

    this.instance = setInterval(() => {
      this.update();
    }, this.options.refreshRate);
  }

  /**
   * Stop and clear the output process.
   */
  stop() {
    clearInterval(this.instance);

    logUpdate.clear();
  }

  /**
   * Update and flush the output if the loader is defined.
   */
  update() {
    const output = this.render();

    if (output) {
      logUpdate(output);
    }
  }
}
