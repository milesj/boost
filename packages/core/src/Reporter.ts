/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import chalk from 'chalk';
import Console from './Console';
import Module from './Module';
import ModuleLoader from './ModuleLoader';
import Task from './Task';
import themePalettes from './themes';
import { Color, ColorType, ColorPalette } from './types';

export const SLOW_THRESHOLD = 10000; // ms

export default class Reporter<Line = string, Options = {}> extends Module<Options> {
  // @ts-ignore Set after instantiation
  console: Console;

  lines: Line[] = [];

  startTime: number = 0;

  stopTime: number = 0;

  /**
   * Register console listeners.
   */
  bootstrap() {
    this.console.on('start', this.handleBaseStart).on('stop', this.handleBaseStop);
  }

  /**
   * Add a line to be rendered.
   */
  addLine(line: Line): this {
    this.lines.push(line);

    return this;
  }

  /**
   * Display an error and it's stack.
   */
  displayError(error: Error): void {
    this.console.write(`\n${this.style(error.message, 'failure', ['bold'])}\n`);

    // Remove message line from stack
    if (error.stack) {
      const stack = this.style(
        error.stack
          .split('\n')
          .slice(1)
          .join('\n')
          .trim(),
        'pending',
      );

      this.console.write(stack, 1);
    }

    this.console.write('\n');
  }

  /**
   * Find a line using a callback
   */
  findLine(callback: (item: Line) => boolean): Line | undefined {
    return this.lines.find(line => callback(line));
  }

  /**
   * Return specific colors based on chosen theme.
   */
  getColorPalette(): ColorPalette {
    const { theme } = this.console.options;
    let palette = {};

    if (chalk.level >= 2) {
      if (themePalettes[theme]) {
        palette = themePalettes[theme];
      } else if (theme !== 'default') {
        palette = new ModuleLoader<ColorPalette>(this.console.tool, 'theme', null, true).loadModule(
          theme,
        );
      }
    }

    return {
      default: 'white',
      failure: 'red',
      pending: 'gray',
      success: 'green',
      warning: 'yellow',
      ...palette,
    };
  }

  /**
   * Return a specific color for each task status.
   */
  getColorType(task: Task<any>): ColorType {
    if (task.isSkipped()) {
      return 'warning';
    } else if (task.hasPassed()) {
      return 'success';
    } else if (task.hasFailed()) {
      return 'failure';
    }

    return 'pending';
  }

  /**
   * Calculate the elapsed time and highlight as red if over the threshold.
   */
  getElapsedTime(start: number, stop: number = 0, highlight: boolean = true): string {
    const time = (stop || Date.now()) - start;
    const isSlow = time > SLOW_THRESHOLD;
    const elapsed = `${(time / 1000).toFixed(2)}s`; // eslint-disable-line no-magic-numbers

    return isSlow && highlight ? this.style(elapsed, 'failure') : elapsed;
  }

  /**
   * Set start time.
   */
  handleBaseStart = () => {
    this.startTime = Date.now();
  };

  /**
   * Set stop time and render.
   */
  handleBaseStop = () => {
    this.stopTime = Date.now();
  };

  /**
   * Create an indentation based on the defined length.
   */
  indent(length: number = 0): string {
    return ' '.repeat(length);
  }

  /**
   * Remove a line to be rendered.
   */
  removeLine(callback: (line: Line) => boolean): this {
    this.lines = this.lines.filter(line => !callback(line));

    return this;
  }

  /**
   * Create a chalk formatted string with accessible colors and modifiers applied.
   */
  style(
    message: string,
    type: ColorType = 'default',
    modifiers: ('reset' | 'bold' | 'dim' | 'italic' | 'underline' | 'inverse' | 'hidden')[] = [],
  ): string {
    const color = this.getColorPalette()[type];
    let out = color.charAt(0) === '#' ? chalk.hex(color) : chalk[color as Color];

    modifiers.forEach(modifier => {
      out = out[modifier];
    });

    return out(message);
  }
}
