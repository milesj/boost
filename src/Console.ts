/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import chalk from 'chalk';
import optimal, { bool, string } from 'optimal';
import ExitError from './ExitError';
import { TaskInterface } from './Task';
import { ReporterInterface } from './Reporter';
import { ConsoleOptions, Partial } from './types';

export interface ConsoleInterface {
  options: ConsoleOptions;
  reporter: ReporterInterface;
  debug(message: string): void;
  error(message: string): void;
  exit(message: string | Error | null, code: number): void;
  log(message: string): void;
  start(tasks: TaskInterface[]): void;
  startDebugGroup(group: string): void;
  stopDebugGroup(): void;
  update(): void;
}

export const DEBUG_COLORS: string[] = [
  'white',
  'cyan',
  'blue',
  'magenta',
  'red',
  'yellow',
  'green',
];

export default class Console<Tr extends ReporterInterface> implements ConsoleInterface {
  debugs: string[] = [];

  debugGroups: string[] = [];

  debugIndex: number = -1;

  errors: string[] = [];

  interrupted: boolean = false;

  logs: string[] = [];

  options: ConsoleOptions;

  reporter: Tr;

  constructor(reporter: Tr, options: Partial<ConsoleOptions> = {}) {
    this.reporter = reporter;
    this.options = optimal(options, {
      debug: bool(),
      footer: string().empty(),
      header: string().empty(),
      silent: bool(),
    });

    // Avoid binding listeners while testing
    if (process.env.NODE_ENV === 'test') {
      return;
    }

    // Start early so we may capture uncaught/unhandled
    this.start();

    const signalHandler = () => {
      if (this.interrupted) {
        this.exit('Process has been terminated.');
      } else {
        this.log(chalk.yellow('Press Ctrl+C again to exit.'));
        this.interrupted = true;
      }
    };

    process
      .on('SIGINT', signalHandler)
      .on('SIGTERM', signalHandler)
      .on('uncaughtException', error => {
        this.debug(chalk.yellow('Uncaught exception detected!'));
        this.exit(error);
      })
      .on('unhandledRejection', error => {
        this.debug(chalk.yellow('Unhandled promise rejection detected!'));
        this.exit(error);
      });
  }

  /**
   * Add a message to the debug log.
   */
  debug(message: string) {
    const indent = '  '.repeat(this.debugGroups.length);

    this.debugs.push(`${chalk.gray('[debug]')}${indent} ${message}`);
  }

  /**
   * Add a message to the error log.
   */
  error(message: string) {
    this.errors.push(message);
  }

  /**
   * Force exit the application.
   */
  /* istanbul ignore next */
  exit(message: string | Error | null, code: number = 1) {
    let errorCode = code;

    // Null messages are always successful
    if (message !== null) {
      const error = message instanceof Error ? message : new ExitError(message, code);

      if (error instanceof ExitError && typeof error.code === 'number') {
        errorCode = error.code;
      }

      // Be sure to output the stack if it exists.
      // Also, we should only show this if there are no error messages,
      // else commands that threw an error but also wrote to stderr,
      // would show twice the output. This is tricky to handle.
      if (this.errors.length === 0) {
        this.error(chalk.red(error.stack || error.message));
      }
    }

    // Stop the renderer
    this.stop();

    // Send final output and exit after buffer is flushed
    (errorCode === 0 ? process.stdout : process.stderr).write(
      this.reporter.render(errorCode),
      () => {
        process.exitCode = errorCode;
      },
    );
  }

  /**
   * Add a message to the output log.
   */
  log(message: string) {
    this.logs.push(message);
  }

  /**
   * Start a new console and begin rendering.
   */
  start(tasks: TaskInterface[] = []) {
    const { debugs, errors, logs } = this;

    this.reporter.start(() => ({
      ...this.options,
      debugs,
      errors,
      logs,
      tasks,
    }));
  }

  /**
   * Start a debug capturing group, which will indent all incoming debug messages.
   */
  startDebugGroup(group: string) {
    this.debugIndex += 1;

    if (this.debugIndex === DEBUG_COLORS.length) {
      this.debugIndex = 0;
    }

    const color = DEBUG_COLORS[this.debugIndex];

    this.debug(chalk[color](`[${group}]`));
    this.debugGroups.push(group);
  }

  /**
   * Stop rendering and end the console.
   */
  stop() {
    this.reporter.stop();
  }

  /**
   * End the current debug capturing group.
   */
  stopDebugGroup() {
    const color = DEBUG_COLORS[this.debugIndex];
    const group = this.debugGroups.pop();

    this.debug(chalk[color](`[/${group}]`));
    this.debugIndex -= 1;

    if (this.debugIndex < 0) {
      this.debugIndex = DEBUG_COLORS.length - 1;
    }
  }

  /**
   * Force a rendering update.
   */
  update() {
    this.reporter.update();
  }
}
