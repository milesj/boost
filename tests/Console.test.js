import chalk from 'chalk';
import Console from '../src/Console';
import Reporter from '../src/Reporter';
import Task from '../src/Task';

describe('Console', () => {
  let cli;

  beforeEach(() => {
    cli = new Console(new Reporter());
  });

  describe('log()', () => {
    it('adds to output log', () => {
      cli.log('foo');

      expect(cli.logs).toEqual(['foo']);
    });

    it('supports formatting', () => {
      cli.log('%s', 'foo');

      expect(cli.logs).toEqual(['foo']);
    });

    it('handles objects and arrays', () => {
      cli.log('%o %O', { foo: 'bar' }, ['baz', 123]);

      expect(cli.logs).toMatchSnapshot();
    });
  });

  describe('error()', () => {
    it('adds to error log', () => {
      cli.error('foo');

      expect(cli.errors).toEqual(['foo']);
    });

    it('supports formatting', () => {
      cli.error('%s', 'foo');

      expect(cli.errors).toEqual(['foo']);
    });

    it('handles objects and arrays', () => {
      cli.error('%o %O', { foo: 'bar' }, ['baz', 123]);

      expect(cli.errors).toMatchSnapshot();
    });
  });

  describe('start()', () => {
    it('calls start on the reporter', () => {
      const spy = jest.fn();
      const task = new Task('Foo', () => {});

      cli.reporter.start = spy;
      cli.start({}, {}, [task]);

      expect(spy).toHaveBeenCalled();
    });

    it('passes value from tool and console', () => {
      jest.useFakeTimers();

      const task = new Task('Foo', () => {});

      cli.options = {
        footer: 'footer',
        header: 'header',
        silent: true,
      };

      cli.log('log');
      cli.error('error');
      cli.start([task]);

      expect(cli.reporter.loader()).toEqual({
        errors: ['error'],
        footer: 'footer',
        header: 'header',
        logs: ['log'],
        silent: true,
        tasks: [task],
      });

      jest.useRealTimers();
    });
  });

  describe('stop()', () => {
    it('calls stop on the reporter', () => {
      const spy = jest.fn();

      cli.reporter.stop = spy;
      cli.stop();

      expect(spy).toHaveBeenCalled();
    });
  });

  describe('update()', () => {
    it('calls update on the reporter', () => {
      const spy = jest.fn();

      cli.reporter.update = spy;
      cli.update();

      expect(spy).toHaveBeenCalled();
    });
  });
});
