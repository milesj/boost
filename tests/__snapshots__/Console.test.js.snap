// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Console error() handles objects and arrays 1`] = `
Array [
  "{ foo: 'bar' } [ 'baz', 123 ]",
]
`;

exports[`Console log() handles objects and arrays 1`] = `
Array [
  "{ foo: 'bar' } [ 'baz', 123 ]",
]
`;
